dashboard/modules/job/job_manager.py

```diff
@@ -288,18 +284,93 @@ class JobManager:
     # Time that we will sleep while tailing logs if no new log line is
     # available.
     LOG_TAIL_SLEEP_S = 1
+    JOB_MONITOR_LOOP_PERIOD_S = 1
 
     def __init__(self):
         self._status_client = JobStatusStorageClient()
         self._log_client = JobLogStorageClient()
         self._supervisor_actor_cls = ray.remote(JobSupervisor)
 
+        self._running_jobs: Dict[str, JobStatusInfo] = dict()
+        self._recover_running_jobs()
+
+    def _recover_running_jobs(self):
+        """Recovers all running jobs from the status client.
+
+        Each will be added to self._running_jobs and reconciled.
+        """
+        all_jobs = self._status_client.get_all_jobs()
+        for job_id, status_info in all_jobs.items():
+            if not status_info.status.is_terminal():
+                self._running_jobs[job_id] = status_info
+                create_task(self._monitor_job(job_id))
+
     def _get_actor_for_job(self, job_id: str) -> Optional[ActorHandle]:
         try:
             return ray.get_actor(self.JOB_ACTOR_NAME.format(job_id=job_id))
         except ValueError:  # Ray returns ValueError for nonexistent actor.
             return None
 
+    async def _monitor_job(
+        self, job_id: str, job_supervisor: Optional[ActorHandle] = None
+    ):
+        """Monitors the specified job until it enters a terminal state.
+
+        This is necessary because we need to handle the case where the
+        JobSupervisor dies unexpectedly.
+        """
+        is_alive = True
+        if job_supervisor is None:
+            job_supervisor = self._get_actor_for_job(job_id)
+
+            if job_supervisor is None:
+                logger.error(f"Failed to get job supervisor for job {job_id}.")
+                self._status_client.put_status(
+                    job_id,
+                    JobStatusInfo(
+                        status=JobStatus.FAILED,
+                        message=(
+                            "Unexpected error occurred: Failed to get job supervisor."
+                        ),
+                    ),
+                )
+                is_alive = False
+
+        while is_alive:
+            try:
+                await job_supervisor.ping.remote()
+                await asyncio.sleep(self.JOB_MONITOR_LOOP_PERIOD_S)
+            except Exception as e:
+                is_alive = False
+                if self._status_client.get_status(job_id).status.is_terminal():
+                    # If the job is already in a terminal state, then the actor
+                    # exiting is expected.
+                    pass
+                elif isinstance(e, RuntimeEnvSetupError):
+                    logger.info(f"Failed to set up runtime_env for job {job_id}.")
```

Comment:
Why `info` and not `error` here?  (I know this isn't your change, but just curious)

Reply:
ah, this was my code originally too :)

this is `INFO` because it is a user error, not an error in job submission, and these are system-level logs. from the standpoint of the job submission server, this is a standard thing that can happen